Hex v2.5 blog post (#1664)
hexpm/hexpm: lib/hexpm_web/templates/blog/023-hex-v25-released.html.md

This release adds three layers of defense. Security advisories are now surfaced directly in your terminal when you fetch or update dependencies. A new release-age cooldown keeps recently released versions out of resolution until they have had time to be vetted. And organizations can publish dependency policies that enforce these rules centrally across every project or user that opts in.

### Security advisories in your terminal

Hex has long warned about retired packages. It now does the same for packages with known security advisories. When you run `mix deps.get` or `mix deps.update` and a resolved version carries an advisory, Hex flags it inline and prints a summary at the end of the run.

<img src="/images/blog/023_advisory.png" srcset="/images/blog/023_advisory.png 2x" alt="mix deps.get showing a VULNERABLE tag and an advisory summary">

This requires no configuration and applies to every project. Advisories that are aliased across multiple databases are now deduplicated, so a single vulnerability is reported once rather than several times.

> The `mix deps.get` and `mix deps.update` warnings are informational and do not fail the build. For a hard check, run `mix hex.audit`. It inspects your lockfile and exits with a non-zero status if any dependency carries an advisory or has been retired, which makes it a natural step to add to CI.

Reviewer: Maybe we should add a line that `mix deps.get` and `mix deps.update` will break the build if you use policies and there is a new dependencies that violates the policy. Perhaps:

```suggestion
The `mix deps.get` and `mix deps.update` warnings are informational and do not fail the build by default. You can make them stricter in two ways:

  * Set a policy and/or cooldown periods that will restrict adding any new dependencies during `mix deps.get` or `mix deps.update`

  * Run `mix hex.audit` to inspect your lockfile and exit with a non-zero status if any existing dependency carries an advisory or has been retired, which makes it a natural step to add to CI
```

When the cooldown hides a version, Hex tells you which ones and when they become eligible.

<img src="/images/blog/023_cooldown.png" srcset="/images/blog/023_cooldown.png 2x" alt="mix deps.get showing versions filtered by cooldown">

`mix hex.outdated` is cooldown-aware too. Held-back updates are annotated with `(cooldown)` rather than hidden, so you can still see what is available and when it becomes installable. If you operate your own organization repository and want hotfixes to bypass the wait, list it under `cooldown_exclude_repos`.

### Dependency policies

> Cooldowns and advisory rules are useful per project, but on a team you do not want every developer to remember to configure them, and you do not want the rules to drift between repositories. Dependency policies move the rules to where they belong: published once by an organization, enforced everywhere that opts in. This is the feature we previewed when we [released Hex 2.4](/blog/hex-v24-released), aimed at attacks like the [axios compromise on npm](https://www.stepsecurity.io/blog/axios-compromised-on-npm-malicious-versions-drop-remote-access-trojan).

Reviewer 2: What "advisory rules" are we referring to?

Author: Security advisories, we call them just "advisories" in a few places in the UI where I think it's clear from context.

Should we clarify in the blog post?

Reviewer 2: It’s just that “rules” implies more than just showing advisories.

Author: This is under the dependency policy section so it's about enforcing rules, not just showing advisories.

Displaying of advisories in under the "Security advisories in your terminal" section.

Reviewer 2: It sounded to me like instead of just rules and advisory rules, we have policies.

But that might just be me. Feel free to close this.

An organization admin creates a named policy from the hex.pm dashboard. A policy targets one or more repositories, typically the organization's own private repo and the public `hexpm` repo, and for each one it sets a restriction and any number of per-package overrides.

<img class="blog-img-light" src="/images/blog/023_policy_create.png" srcset="/images/blog/023_policy_create.png 2x" alt="Creating a dependency policy in the hex.pm dashboard">
<img class="blog-img-dark" src="/images/blog/023_policy_create_dark.png" srcset="/images/blog/023_policy_create_dark.png 2x" alt="Creating a dependency policy in the hex.pm dashboard">

A restriction can block releases that carry a security advisory at or above a chosen severity, are retired for one of a chosen set of reasons, or are newer than a release-age cooldown window. Overrides handle the exceptions. An allow override exempts a specific package, or a specific version range of it, from the restriction; a deny override blocks one outright.

> Policies are signed and served through the organization's repository, the same trusted channel as the package registry itself, so the client can verify them. Once published, a policy's name is immutable.

Reviewer 2: ```suggestion
Policies are signed and served through the organization's repository, the same trusted channel as the package registry itself, so the client can verify them.
```

The immutable part is less relevant for a blogpost.

```elixir
# mix.exs
def project do
  [
    # ...
    hex: [policy: [org: "myorg", name: "strict-prod"]]
  ]
end
```

You can also use `HEX_POLICY=hexpm:myorg/strict-prod` or `mix hex.config policy hexpm:myorg/strict-prod`. Policy enforcement fails closed: if a policy is configured but cannot be fetched or is malformed, resolution stops rather than proceeding unprotected.

Two new commands make the active policy legible from the terminal. `mix hex.policy show` summarizes it:

<img src="/images/blog/023_policy.png" srcset="/images/blog/023_policy.png 2x" alt="mix hex.policy show output">

And `mix hex.policy why PACKAGE` walks every version of a package and shows exactly which are blocked and why:

<img src="/images/blog/023_policy_why.png" srcset="/images/blog/023_policy_why.png 2x" alt="mix hex.policy why phoenix_live_view output">

During a normal `mix deps.get`, policy filtering is reported alongside the cooldown summary, with a per-version listing of what was hidden. If a restriction blocks every version a direct dependency could resolve to, resolution fails with a focused note explaining which versions the policy hid, so the cause is never a mystery.

> ### How the layers fit together

Reviewer 2: I feel like that’s better for docs than an announcement post.

The three features are independent but reinforce each other, and they compose by taking the strictest rule. A local cooldown and a policy cooldown combine to the longer of the two, and a policy sets a floor that local configuration cannot lower. Locked versions are always trusted, and known-unsafe locked versions always have an escape route. The result is a defense you can adopt incrementally: surface advisories for free, turn on a cooldown for yourself today, and roll out an organization-wide policy when you are ready.

> 

Author: ```suggestion
```
